electrumpersonalserver/server/peertopeer.py

```diff
@@ -0,0 +1,358 @@
+                            net_magic, command, payload_length, checksum = unpack('<I12sI4s', recv_buffer[:HEADER_LENGTH])
+                            recv_buffer = recv_buffer[HEADER_LENGTH:]
+                            if net_magic != self.magic:
+                                log.debug('wrong MAGIC: ' + hex(net_magic))
+                                self.sock.close()
+                                break
+                            command = command.strip(b'\0')
+                        else:
+
+                            if payload_length >= 0 and len(recv_buffer) >= payload_length:
+                                payload = recv_buffer[:payload_length]
+                                recv_buffer = recv_buffer[payload_length:]
+                                if btc.bin_dbl_sha256(payload)[:4] == checksum:
+                                    self.p2p_message_handler.handle_message(self, command,
+                                        payload_length, payload)
+                                else:
+                                    log.debug('wrong checksum, dropping message, cmd=' + command + ' payloadlen=' + str(payload_length))
+                                payload_length = -1
+                                data_remaining = True
+                            else:
+                                data_remaining = False
+                except socket.timeout:
+                    self.p2p_message_handler.check_keepalive(self)
+                    self.p2p_message_handler.on_heartbeat(self)
+        except EOFError as e:
+            self.closed = True
+        except IOError as e:
+            import traceback
+            log.debug("logging traceback from %s: \n" %
+                traceback.format_exc())
+            self.closed = True
+        finally:
+            try:
+                self.sock.close()
+            except Exception as _:
+                pass
+
+
+    def close(self):
+        self.closed = True
+
+    def create_message(self, command, payload):
+        return (pack("<I12sI", self.magic, command.encode(), len(payload))
+            + btc.bin_dbl_sha256(payload)[:4] + payload)
+
+class P2PBroadcastTx(P2PMessageHandler):
+    def __init__(self, txhex):
+        P2PMessageHandler.__init__(self)
+        self.txhex = bytes.fromhex(txhex)
+        self.txid = btc.bin_txhash(self.txhex)
+        log.debug('broadcasting txid ' + str(self.txid))
+        self.uploaded_tx = False
+        self.time_marker = datetime.now()
+        self.connected = False
+
+    def on_recv_version(self, p2p, version, services, timestamp,
+            addr_recv_services, addr_recv_ip, addr_trans_services,
+            addr_trans_ip, addr_trans_port, user_agent, start_height, relay):
+        if not relay:
+            log.debug('peer not accepting unconfirmed txes, trying another')
+            # this happens if the other node is using blockonly=1
+            p2p.close()
+        if not services & NODE_WITNESS:
```

Comment:
`getnodeaddresses` gives you the service bits of each address, so check there before you've even connected instead of here.

Reply:
How does it know what services the addresses have if it's never connected to these nodes? They could be out of date from when we received that information, and now accept witness data.

Reply:
When nodes gossip addresses to each other they also include the service bits (as well as timestamp). So that's how it knows. If a node doesn't accept witnesses there's no point trying to connect given that there are 10k others to choose from.